### ISSUE TITLE
Clash should warn against just using `NOINLINE` for primitives on GHC >= 9.4

### ISSUE THREAD

**Issue by martijnbastiaan** [clash-lang/clash-compiler#2510]
The following:

```haskell
import Data.String.Interpolate (__i)
import Clash.Explicit.Prelude
import Clash.Annotations.Primitive

class X a where
  x :: a

instance X (Signal System Int) where
  x = pure 3

instance X a => X (Signal System Int -> a) where
  x !_i = x

bb :: X a => a
bb = x
{-# NOINLINE bb #-}
{-# ANN bb hasBlackBox #-}
{-# ANN bb (
  let bbName = show 'bb
  in InlineYamlPrimitive [minBound..] [__i|
    BlackBox:
      name: "#{bbName}"
      kind: Expression
      template: "this does not end up in HDL with NOINLINE~ARG[1]"
|]) #-}

bbWrapper :: X a => a
bbWrapper = bb
{-# NOINLINE bbWrapper #-}

topEntity :: Signal System Int -> Signal System Int
topEntity i = bbWrapper i
```

..does not end up using the blackbox of `bb`, but instead uses the core unfolding. Clash point this out and encourage use of `OPAQUE`.
